### PR TITLE
Enable construction of vectors from `std::initializer_list`

### LINKED ISSUE #1835: `thrust::device_vector` should have an `std::initializer_list` constructor
Currently, its not possible to do:

```
thrust::device_vector<int> v{4, 8, 0, 8, 4, 2, 8};
```

But instead we need to do:

```
thrust::device_vector<int> v{std::vector<int>{4, 8, 0, 8, 4, 2, 8}};
```

Because there are no constructor taking a `std::initializer_list`, unlike `std::vector`.

Could it be possible to add this?

This shoud be straightforward to add, using other constructors.

### PR DESCRIPTION
**Pull request by miscco**
This fixes #1835 and enables construction of vectors from an `initializer_list` as well as assignment.

### CODE REVIEW

**Comment by miscco** [comment]
run tests

**Comment by miscco** [comment]
run tests

File: thrust/host_vector.h [hunk 1]
```diff
@@ -25,6 +25,10 @@
 #include <thrust/detail/config.h>
 #include <thrust/detail/memory_wrapper.h>
 #include <thrust/detail/vector_base.h>
+
+#if THRUST_CPP_DIALECT >= 2011
```
Comment (senior-zero) [inline thrust/host_vector.h:29]: I don't think we support CPP < 2011

File: thrust/device_vector.h [hunk 2]
```diff
@@ -197,6 +200,27 @@ template<typename T, typename Alloc = thrust::device_allocator<T> >
     device_vector &operator=(const detail::vector_base<OtherT,OtherAlloc> &v)
     { Parent::operator=(v); return *this; }
 
+#if THRUST_CPP_DIALECT >= 2011
```
Comment (senior-zero) [inline thrust/device_vector.h:203]: I don't think we support CPP < 2011

File: testing/vector.cu [hunk 3]
```diff
@@ -37,6 +40,32 @@ void TestVectorBool(void)
 }
 DECLARE_UNITTEST(TestVectorBool);
 
+#if THRUST_CPP_DIALECT >= 2011
+template <class Vector>
+void TestVectorInitializerList(void)
```
Comment (senior-zero) [inline testing/vector.cu:45]: Optional: it'd be nice to test non-trivial type here as well.

**Review by senior-zero — APPROVED** [review]
Thanks for the addition! A few optional comments below.

File: thrust/device_vector.h [hunk 4]
```diff
@@ -197,6 +200,27 @@ template<typename T, typename Alloc = thrust::device_allocator<T> >
...
+    /*! This constructor builds a \p device_vector from an intializer_list.
+     *  \param il The intializer_list.
+     */
+    device_vector(std::initializer_list<T> il)
+      :Parent(il.begin(), il.end()) {}
```
[the remark this answers is not in the document]
Comment (miscco) [reply thrust/device_vector.h]: Thanks a lot.

**Comment by miscco** [comment]
run tests

File: thrust/detail/vector_base.inl [hunk 5]
```diff
@@ -195,6 +195,34 @@ template<typename T, typename Alloc>
...
 } // end vector_base::operator=()
 
+  template<typename T, typename Alloc>
+    vector_base<T,Alloc>
+      ::vector_base(std::initializer_list<T> il)
+        :m_storage(),
+         m_size(0)
+  {
+    range_init(il.begin(), il.end());
+  } // end vector_base::vector_base()
```
Comment (jrhemstad) [inline thrust/detail/vector_base.inl:205]: I'm lazy, so I'm always a fan of delegating to other constructors when possible. In this case, we can delegate to the iterator pair ctor: https://github.com/NVIDIA/thrust/blob/37a638e54a58a43955281c3c0c76b7627d34f1e6/thrust/detail/vector_base.inl#L309 

```suggestion
  template<typename T, typename Alloc>
    vector_base<T,Alloc>
      ::vector_base(std::initializer_list<T> il) : vector_base(std::cbegin(il), std::cend(il)) {}
```

Comment (miscco) [reply thrust/detail/vector_base.inl:205]: We did not touch those constructors and we should either move all of them to a consistent pattern or leave them as is.

**Pull request merged**